blog/_posts/2025-04-04-Optimize-AKS-Traffic-with-externalTrafficPolicy=Local.md

```diff
@@ -0,0 +1,152 @@
+
+### Preventing New Connections to an Unhealthy Pod when using externalTrafficPolicy=Local
+
+To avoid routing new requests to a pod that is in the process of shutting down, it is important to manage its health status effectively. The below image shows the timeline for a pod receiving a `TERM` signal and gracefully shutting down without impact on external traffic:
+
+![Preventing New Requests to Unhealthy Pods](./AKS/blog/assets/images/preventingnewrequeststoanunhealthypod.png)
+
+- **Immediate Health Check Response**:
+    As soon as a pod is marked for deletion, its `healthCheckNodePort` should start returning a 500 error. This signals to external load balancers that the pod is no longer healthy and should not receive new traffic.
+
+- **Load Balancer Probe Delay**:
+    External load balancers will take a few seconds (upto 10 sec) to detect the unhealthy status of a pod. During this time, the pod might still receive new connections.
+
+To ensure your pods follow a similar timeline to gracefully shutdown, make sure to assess the following:
+
+1. **Delay Termination**:
+     After receiving the `TERM` signal, we recommend your application wait for at least 10 seconds before proceeding with shutdown tasks (to address the Load balancer probe delay). This can be achieved using a `preStop` hook in Kubernetes.
+
+    > **Note**: When using annotations `service.beta.kubernetes.io/azure-load-balancer-health-probe-interval` and/or `service.beta.kubernetes.io/azure-load-balancer-health-probe-num-of-probe`, consider changing the wait time to cover the annotation's needs (see [documentation](https://cloud-provider-azure.sigs.k8s.io/topics/loadbalancer/))
+
+2. **Announce Readiness as `false`**:
+     The application should update its readiness probe to indicate it is no longer ready to serve traffic after the above delay has completed. This allows Kubernetes to stop routing new connections to the pod after the load balancer removes it from the list of active pods.
+
+3. **Gracefully Close Existing Connections**:
+     The application should close all active connections, ensuring that no in-flight requests are dropped.
+
+4. **Exit the Process**:
+     Once all shutdown tasks are complete, the application should terminate its process cleanly.
+
+By implementing these best practices, you can minimize disruptions during pod shutdowns, maintain a seamless user experience, and ensure the reliability of your services in production environments.
+
+## Best Practices for Rolling Updates and Pod Rotation
+![Pod Rotation Timeline](./AKS/blog/assets/images/podrotationbestpractices.png)
+
+While the above works when a pod is being taken down in isolation, it does not cover cases like upgrades and rolling restarts which require coordination between the time the pod goes down and a new one comes up, ready to serve traffic. To optimize pod rotation, add the following best practice to your deployment:
+
```

Comment:
Link to Kubernetes upstream docs on these will be welcome. say on minreadyseconds.